### PR TITLE
gromacs: build several flavors (parallel x precision) in the same pac…

### PR DESCRIPTION
**Pull request by ggouaillardet**
…kage

add two new variants :
 - precision = { single, double, all }
 - parallel = { thread-mpi, mpi, all }
and remove the existing variants :
 - double
 - mpi

End users can easily be confused when several GROMACS package with
the same version and compiler are available, for example when they
are used to do some light pre/post processing with the sequential
version (e.g. gmx ...) and the heavy computation with the MPI version
(e.g. mpirun gmx_mpi ...)

Per the feedback to https://mailman-1.sys.kth.se/pipermail/gromacs.org_gmx-developers/2021-April/011061.html,
this enhancement allows to build several parallelization paradigm (thread-mpi and/or mpi)
and precisions (single vs double) with the same package.
For example
spack install gromacs parallel=all precision=all
will make the four binaries available with a single
spack load gromacs
command :
 - gmx (single precision, thread-mpi)
 - gmx_mpi (single_precision, mpi)
 - gmx_d (double precision, thread-mpi)
 - gmx_mpi_d (double precision, mpi)

### CODE REVIEW

**Comment by ggouaillardet** [comment]
@junghans @marvinbernhardt could you please review this PR?

It removes the `double` and `mpi` variants, but I can "restore" them
if needed in order to kind-of maintain the
`spack install gromacs ...`
command line compatibility.

**Comment by marvinbernhardt** [comment]
I see why this would be nice, having only one gromacs package in spack is convenient.
Three questions:

1. libraries: If I understand correctly each version will have it's own subdirectory. But will then not all versions end up in `$LD_LIBRARY_PATH`? Could the `gmx_d` try to load a mixed precision library? Maybe not because of RPATH. But maybe this is still problematic for other programs that want to use gromacs libraries.
2. spack policy: This way of packaging is taking part of spacks job. I haven't found any policy on this, but I could imagine that “putting multiple variants in one install” can be taken to the extreme and it might be best to not ever start it. @alalazo ?
3. versions: Does this PR work for older gromacs versions (4., 5.)?

Also, I think variants should be changed to [mulit-valued variants](https://spack.readthedocs.io/en/latest/packaging_guide.html#multi-valued-variants).

**Comment by ggouaillardet** [comment]
@marvinbernhardt each version is built (e.g. `cmake && make && make install`) in its own directory, but everything ends up in the same prefix.
That works since each version produces unique binaries and libraries (e.g. `gmx` and `gmx_mpi_d`). Some files in `<prefix>/share` gets overwritten ... with identical files.

Thanks for the feedback, I will shortly push an updated PR that uses multi-valued variants.

File: var/spack/repos/builtin/packages/gromacs/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/gromacs/package.py]: This doesn't make sense to me as a multi-valued variant. I think we should either have it `multi=False` or remove the value `all`.

File: var/spack/repos/builtin/packages/gromacs/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/gromacs/package.py]: Same comment here.

**Review by alalazo — CHANGES_REQUESTED** [review]
I left a couple of technical comments on this change below. More in general I wonder what would be the practical gain with this formulation, which seems to make the package recipe more complex?

File: var/spack/repos/builtin/packages/gromacs/package.py
Comment (ggouaillardet) [reply var/spack/repos/builtin/packages/gromacs/package.py]: Good catch, I updated the commit message but forgot to remove this from the code (!)

**Comment by ggouaillardet** [comment]
@alalazo yes, this PR does make the recipe more complex.

The point is it makes enduser experience much simpler and less error-prone:

Today, it is quite common to
```
$ spack load gromacs -mpi
$ gmx ...                 # sequential pre-processing
$ spack unload gromacs
$ spack load gromacs +mpi
$ mpirun gmx_mpi ...      # parallel computation
$ spack unload gromacs
$ spack load gromacs -mpi
$ gmx ...                 # sequential post-processing
```

with this PR, it becomes
```
$ spack load gromacs      # no more need to specify variants
$ gmx ...                 # sequential pre-processing
$ mpirun gmx_mpi ...      # parallel computation
$ gmx ...                 # sequential post-processing
```

**Comment by alalazo** [comment]
> The point is it makes enduser experience much simpler and less error-prone

If that case is predominant why maintaining the variants and not trying to install everything every time? (Disclaimer: I am not a `gromacs` user so I don't know which workflow are common / uncommon. I am just wondering why doing something that works around what the underlying build-system currently permits). 

I'll keep the request for change until maintainers have reviewed and approved the change. Ping me if I am the only factor holding this up

**Comment by ggouaillardet** [comment]
@alazo I should have written

> The point is to give sys/spack admins the option to make enduser life simpler

GROMACS takes some time to build, so unconditionally building 4 flavors instead of one does take a while.

By using variants, sys/spack admins have the flexibility to build one package per flavor (just like today), or one package with up to 4 flavors.

Thanks for the feedback !

**Comment by junghans** [comment]
I am not a 100% convinced this is useful, the mpi-enabled binaries can be run in serial, too.

Needing multiple precisions at the same time is even less likely.

But maybe I am not seeing the use case here....

**Comment by marvinbernhardt** [comment]
> I am not a 100% convinced this is useful, the mpi-enabled binaries can be run in serial, too.
> 
> Needing multiple precisions at the same time is even less likely.
> 
> But maybe I am not seeing the use case here....

I was about to write something similar. I don't have much expirience with the gromacs+mpi, but I would be surprised, if one could not just do `gmx_mpi rdf ...`!?

For mixed/double precision, I think the most likely use case is normal mode calculation. But this seems not an everyday thing, and keeping a second version for that task should be fine.

**Comment by ggouaillardet** [comment]
At least one MPI implementation does not support singleton mode, so the only way is to `mpiexec -n 1 gmx_mpi ...`.

I made this PR since I already received more than one ticket from end users who are experts in their field,
and their field is not Spack:
 - why does `spack load gromacs@2021.1` fails?
 - I loaded one gromacs module, I can `mpiexec gmx_mpi`, but where is `gmx`?

I initially requested feedback from the GROMACS developers about installing both serial and MPI flavors in the same prefix,
and the reply was basically "it works, we do this ... and with both single and double precision".
That's why I added this option (both single and double precision) even if I honestly do not see a strong use-case for that.

**Comment by marvinbernhardt** [comment]
> At least one MPI implementation does not support singleton mode, so the only way is to `mpiexec -n 1 gmx_mpi ...`.

This does indeed look a bit ugly, but managable.

>     * why does `spack load gromacs@2021.1` fails?

I would say Spack gives pretty good error messages in that case.

>     * I loaded one gromacs module, I can `mpiexec gmx_mpi`, but where is `gmx`?

Well, I would just tell the users to use gmx_mpi for everything.

Overall, I see the small convenience improvement as a plus. But in the long run, I think it is not a good idea to start merging variants into one installation with different suffixes. So in the end I am against merging.

**Comment by tz-rrze** [comment]
I like the idea of the pull request and can confirm that Gromacs users expect such a behavior and that they switch between serial / thread-mpi / mpi for different phases of a project. (However, I have not seen them switching between single and double precision.)
Generally using ``gmx_mpi`` is a bad idea. For single node execution - even with multiple processes - the thread-mpi version of Gromacs (which is included in the non-mpi version) is typically significantly faster. It's even more important for the GPU versions ...

**Comment by marvinbernhardt** [comment]
> I like the idea of the pull request and can confirm that Gromacs users expect such a behavior and that they switch between serial / thread-mpi / mpi for different phases of a project. (However, I have not seen them switching between single and double precision.)
> Generally using `gmx_mpi` is a bad idea. For single node execution - even with multiple processes - the thread-mpi version of Gromacs (which is included in the non-mpi version) is typically significantly faster. It's even more important for the GPU versions ...

That seems to be true: using gmx_mpi on a single node is not ideal. But from my perspective it very straightforward to tell users to `spack load gromacs+mpi` for large systems on multiple nodes and `spack load gromacs~mpi` for small systems running on single nodes.

**Comment by haampie** [comment]
There's also interest from @eth-cscs to have one Gromacs build with all the flavors (it's how it's built with Easybuild). My suggestion is to make Gromacs support this in their build system instead of hacking the Spack package to build multiple builds -- they already generate executables with different names depending on cmake config options, how hard can it be to just generate more than 1 executable. Are there are any Gromacs devs in this thread who can comment if this is feasible? (ping @ggouaillardet, @marvinbernhardt, @junghans, @tz-rrze)

**Comment by victorusu** [comment]
The main use cases for @eth-cscs to build multiple variants of GROMACS in the same module are
1. one cannot run MPI-enabled applications on Cray login nodes. So [if there are multiple modules as in the current Spack] the users have to load different modules to create input files compared to the one used for running simulations. This created too many tickets, so we stopped doing that.
2. Furthermore, in general, the GROMACS module visible immediately in a Hierarchical naming scheme is the non-MPI one, which can be misleading to users. 
3. There are some specific simulations/workflows that require double-precision, which implies again in another set of modules. We had an actual ticket where the system explodes when using single-precision due to the fact that the box is very, very elongated. So, we also provide a double-precision version.
4. Depending on the input file and on the hardware used (I mean accelerators like GPGPUs) running simulations with `gmx_mpi` can be more efficient in a single node compared to `gmx`. The "default" regression test that we run on our flag system every day runs faster with `gmx_mpi` compared to `gmx` in a single node - we use [12 tasks per node because of that](https://github.com/eth-cscs/reframe/blob/164bae6084dd9590f232f99f5cbeb0beed7ace26/cscs-checks/apps/gromacs/gromacs_check.py#L71).

**Comment by junghans** [comment]
It still don't see the multi-precision use case as one should usually not mix precision within one simulation setup.

The use case with `gmx` and `gmx_mpi` I kind of understand, also because e.g. `votca-csg` is using the serial `libgromacs` to read files, but can use the parallel `gmx_mpi` to run simulation at the same time. And spack currently makes it impossible to support that.

In conclusion as long as one of you steps up as maintainer for gromacs, I am ok to merge this.

**Comment by ggouaillardet** [comment]
I am happy to maintain the multi build parts of the GROMACS recipe.

As far as I am concerned, I do not need multi-precision (I basically did it because I could), so let's try to reach a consensus to decide what we are going to implement.

I will also update the PR since I was able to simplify it (e.g. use `super()` instead of copy/paste bunch of code).

And then this PR will be ready for prime time.

**Comment by alalazo** [comment]
> how hard can it be to just generate more than 1 executable. Are there are any Gromacs devs in this thread who can comment if this is feasible? 

Anybody who can comment on this :point_up: ?

**Comment by junghans** [comment]
> > how hard can it be to just generate more than 1 executable. Are there are any Gromacs devs in this thread who can comment if this is feasible?
> 
> Anybody who can comment on this ☝️ ?

As a gromacs dev, I would say not going to happen anytime soon, but @acmnpv and @mabraham @eirrgang might have a different opinions.

**Review by junghans — DISMISSED** [review]

**Comment by ejjordan** [comment]
> > how hard can it be to just generate more than 1 executable. Are there are any Gromacs devs in this thread who can comment if this is feasible?
> 
> Anybody who can comment on this  ?

I think I would count as a core gromacs dev, and I will say that this is something we would like to support but is not on our roadmap. For combined single/double builds, we would need to template many of the functions in the library, so that single vs double could be decided by the calling code. This is unlikely to happen anytime soon. For supporting thread-MPI and MPI, the amount of work would be slightly less (or at least less tedious), but would still require significant reworking of some back-ends. Currently, most of our effort is going into unifying the various GPU back-ends, which still would not necessarily mean we have a straightforward path to supporting multiple architectures in a singel build, even though we would like to.

**Comment by marvinbernhardt** [comment]
I see there is some request from users, so probably this should get merged. 

Having several installations in one spec (even if they don't conflict yet)  feels like it will lead to problems in some years. So I want to point out an alternative: `spack load` already has `--first` so it could as well have `--all` and that should (could) be easy enough for the users.

**Comment by haampie** [comment]
> So I want to point out an alternative: spack load already has --first so it could as well have --all and that should (could) be easy enough for the users.

Hm, it's better to use environments instead of introducing more flags to spack load, since that's much more transparent. Loading 'all' can load multiple version etc. Also note that spack can be used to only build things for users who expect to do module load gromacs and have no clue about spack.

> For combined single/double builds, we would need to template many of the functions in the library, so that single vs double could be decided by the calling code.

Is this really the case? The goal is not to get one library/executable with double/float mpi/non-mpi etc, but rather to have multiple executables with different suffixes, just like you get if you run cmake-make-install multiple times with the same install prefix like this PR is doing. Can't you make the options you have in CMake multi-valued and then add multiple targets with add_executable in a loop over the values, and then set the compile-time defines appropriately for each target?

**Comment by mabraham** [comment]
> The goal is not to get one library/executable with double/float mpi/non-mpi etc, but rather to have multiple executables with different suffixes, just like you get if you run cmake-make-install multiple times with the same install prefix like this PR is doing. Can't you make the options you have in CMake multi-valued and then add multiple targets with add_executable in a loop over the values, and then set the compile-time defines appropriately for each target?

Yes, one can push down the loops over build types into the CMake and produce different defines and `config.h` files and recompile all the source files to produce multiple executables and libraries. That would make spack's problem easier (and perhaps other projects' problems also) but would take effort away from the things the GROMACS developers are paid for without providing much direct value to GROMACS users or developers. The biggest value to GROMACS developers would be being able to compile and test many library flavors within one build tree, but in the short term the benefit won't be large because each library will need to recompile every source file.

There's certainly some files that should not need to be recompiled if it was done well, but not lots of them unless the pattern of "one `config.h` and one set of compiler flags to rule them all" was broken up. Unfortunately the CMake code in GROMACS was written long before "modern CMake" with target-specific flags and defines. For example, currently all the short-range kernels would be recompiled for the different MPI flavors, even though they don't actually care about MPI at all, because they depend on other things in `config.h`. Similarly there's a lot of code that cares about MPI but not about precision or GPU support. To manage that in a sane way means you want modules within the libraries to get only the configuration settings that pertain to them, which you can only do if you refactor the build system into that "modern CMake" style with lots of targets. That would bring other advantages, and we did some preliminary work for it last year but it didn't get prioritised by me this year...

Once that CMake refactoring was done, there'd be more work updating the tests so that it was clear which configuration(s) had a problem.

**Comment by victorusu** [comment]
Based on the replies from @mabraham @ejjordan, I think this PR should be merged. We can take the complexity now and when GROMACS does the change we revert back to a simpler style. Furthermore, @ggouaillardet has volunteered to maintain this part and I can happily help him to maintain it too. So, I only see win-win situations and more engagement from a larger group of people.

**Comment by junghans** [comment]
@alalazo please review again!

**Comment by junghans** [comment]
@ggouaillardet can you rebase this and add the most recent gromacs releases?

**Comment by ggouaillardet** [comment]
@junghans Done!

**Review by junghans — APPROVED** [review]

**Review by haampie — CHANGES_REQUESTED** [review]
You should fix the `mpi` variant properly in this package, and also fix dependents of gromacs and references to it in documentation.

File: var/spack/repos/builtin/packages/gromacs/package.py
Comment (marvinbernhardt) [inline var/spack/repos/builtin/packages/gromacs/package.py:72]: precision should probably have the option 'mixed' instead of 'single', since this is how Gromacs calls it since at least 2018

**Comment by haampie** [comment]
Can you also share what the `lib/pkgconfig` dir looks like after this change?

I'm still convinced that gromacs should improve their build system instead of forcing this problem on users.

File: var/spack/repos/builtin/packages/gromacs/package.py
Comment (haampie) [reply var/spack/repos/builtin/packages/gromacs/package.py:72]: I don't think that's a good idea because this is a multi-valued variant.

**Comment by eirrgang** [comment]
> You should fix the `mpi` variant properly in this package, and also fix dependents of gromacs and references to it in documentation.

Do you mean that the `mpi` variant should be retained in addition to the new variants? I can see that its removal could be a problem for dependent projects, and I expect many recipes may be maintained outside of this repository.

Would it be reasonable and practical to let `+mpi` imply `parallel=mpi` in the near term and issue a deprecation warning? (Sorry if I'm somewhat inexperienced with Spack.)

**Comment by haampie** [comment]
It's fine to break packages as they are not backported. I'm talking about the whole section in gromacs/package.py referring to `+mpi` and `~mpi`, that should use `parallel=...`.

File: var/spack/repos/builtin/packages/gromacs/package.py
Comment (marvinbernhardt) [reply var/spack/repos/builtin/packages/gromacs/package.py:72]: I don't understand. My proposal was just to rename single to mixed to match the Gromacs terminology.

**Comment by eirrgang** [comment]
> It's fine to break packages as they are not backported. I'm talking about the whole section in gromacs/package.py referring to `+mpi` and `~mpi`, that should use `parallel=...`.

Ah. You mean like `depends_on('plumed~mpi', when='+plumed~mpi')`? This change breaks `when='+plumed~mpi'` checks, right?

File: var/spack/repos/builtin/packages/gromacs/package.py
Comment (haampie) [reply var/spack/repos/builtin/packages/gromacs/package.py:72]: Ok, I thought you meant mixed as equivalent to precision=single,double. I think it's only extra confusing when we allow multiple precision in one build.

**Comment by haampie** [comment]
Can you forward the mpi variants just once and wrap the plumed business in `when(...)`:

```python
with when('+plumed'):
    depends_on('plumed~mpi', when='parallel=thread-mpi')
    depends_on('plumed+mpi', when='parallel=mpi')

    # just versions compat from here
    depends_on('plumed@2.7.1:2.7.2', when='@2021')
    ...
```

Actually, this is not going to work like before because Spack has no means to express a negative match in a `when` statement. So when `parallel=thread-mpi,mpi`, you don't want to depend on both `plumed~mpi` and `plumed+mpi`. The only thing you can do is:

```python
with when('+plumed'):
    depends_on('plumed')
    depends_on('plumed+mpi', when='parallel=mpi')
```

**Comment by haampie** [comment]
Note that if this change will overwrite the same `pkgconfig` file for all flavours, I would still remain unconvinced of this approach and vote against. It seems gromacs dose not generate a cmake config script, but I would have the same concerns with that of course.

If this is a common use case, it should be fixed upstream, and not fixed ad-hoc in Spack.

**Comment by eirrgang** [comment]
> It seems gromacs dose not generate a cmake config script

gromacs installs `share/cmake/gromacs$SUFFIX/gromacs$SUFFIX-config{,-version}.cmake`, etc. Is that what you mean? It may not be obvious from browsing the project CMakeLists.txt files, since the scripting to generate these predates the modern CMake helper functions.

Unfortunately, these do not provide complete support to convey GROMACS dependencies to client software. There may be some momentum building to improve the situation, but the results would not appear before GROMACS 2023.

Note that the CMakeToolchain file that was provided for a few years has been abandoned in favor of a "hints" file (to be provided with `cmake -C ...`) for GROMACS 2022. Feedback would be welcome at https://gitlab.com/gromacs/gromacs/-/issues or http://www.gromacs.org/Support/GMX-Developers_List regarding additions to better hint client software. (hints for `FindMPI.cmake` are certainly on the short list)

**Comment by haampie** [comment]
Ok, my double-version of gromacs finished building and it does have a proper suffix for both pkgconfig and cmake config files, so I don't think I'm opposing this as much anymore.

The only downside right now is that with `threads=mpi,thread-mpi` we do end up with the thread-mpi version having rpaths to the mpi prefix dir. But I guess that's not a blocker.

**Comment by eirrgang** [comment]
> with `threads=mpi,thread-mpi` we do end up with the thread-mpi version having rpaths to the mpi prefix dir. But I guess that's not a blocker.

Ew! That's ugly. I wonder why.

Does `super().build(spec, prefix)` maybe re-use the build directory? Is there a way to explicitly clean it?

The GROMACS CMake infrastructure is "optimized" for developers rebuilding with minor tweaks, and not <s>very</s> resilient to re-running CMake with the intention of a different build configuration. (The use case of multiple overlapping installations (or multiple build configurations in the same package) is not a core targeted use case, and is not tested or designed around.)

Or is this related to the path rewriting that Spack can do to convert build-time paths to install-time paths?

**Comment by haampie** [comment]
Can you verify it? It seems what you have is correct, I haven't tried to build it because it doesn't concretize without the +/~mpi fixes.

The rpath to mpi is there because CMakePackage sets `CMAKE_INSTALL_RPATH` to the package's link-type deps.

**Comment by eirrgang** [comment]
> The rpath to mpi is there because CMakePackage sets `CMAKE_INSTALL_RPATH` to the package's link-type deps.

Oh, you mean the rpath includes the path for the mpi library dependency. Not that it is resolving to the wrong libgromacs. That doesn't seem like a big problem, but it indicates something is being misused somewhere or assumptions are being violated, so it's worth understanding.

I can't think of how that would happen unless the build directory were being reused for all of the flavors, in which case I can't think of how to (reliably) *prevent* it from happening. (There are too many cache variables that don't get recalculated once set.) Furthermore, it's conceivable that other aspects of the build could end up wrong in subtle ways... (Just speaking from experience, I *highly* recommend a clean build directory for every git ref and combination of cmake options.)

Personally, I would favor discrete `gromacs-nompi`, `gromacs-mpi`, and `gromacs-threadmpi` recipes to reflect how the software is instrumented and (assumed to be) used, pending gromacs decisions to rearchitect for a more unified user experience. It seems to me like separate spack environments or discrete non-conflicting spack packages are both preferable to this solution. If further build troubleshooting is needed, though, I can try to scavenge some time to run the permutations of the spack recipes here, but probably not before December.

**Comment by haampie** [comment]
The `build_directory` as a property trick looks fine (new build dir for every flavor). It's that Spack _always_ passes `CMAKE_INSTALL_RPATH=...` as a default argument when you invoke it from the base class. Again, I don't think this is a major issue.

**Comment by eirrgang** [comment]
> The `build_directory` as a property trick looks fine (new build dir for every flavor). It's that Spack _always_ passes `CMAKE_INSTALL_RPATH=...` as a default argument when you invoke it from the base class. Again, I don't think this is a major issue.

Ah. So it's *Spack* that considers the mpi-implied rpath to be a characteristic of all of the cmake projects configured by a single instance of `CMakePackage`. That seems like a sensible assumption, and implies that recipes like this are non-normative. It is what it is, I guess, and not a major issue.

**Comment by marvinbernhardt** [comment]
> Can you forward the mpi variants just once and wrap the plumed business in `when(...)`:
> 
> ```python
> with when('+plumed'):
>     depends_on('plumed~mpi', when='parallel=thread-mpi')
>     depends_on('plumed+mpi', when='parallel=mpi')
> 
>     # just versions compat from here
>     depends_on('plumed@2.7.1:2.7.2', when='@2021')
>     ...
> ```
> 
> Actually, this is not going to work like before because Spack has no means to express a negative match in a `when` statement. So when `parallel=thread-mpi,mpi`, you don't want to depend on both `plumed~mpi` and `plumed+mpi`. The only thing you can do is:
> 
> ```python
> with when('+plumed'):
>     depends_on('plumed')
>     depends_on('plumed+mpi', when='parallel=mpi')
> ```

This has not been fixed yet.

File: var/spack/repos/builtin/packages/gromacs/package.py
Comment (marvinbernhardt) [inline var/spack/repos/builtin/packages/gromacs/package.py:447]: unpacking of list not needed here, just use product(precisions, parallels)

**Comment by junghans** [comment]
Ping on this feature.

**Comment by ggouaillardet** [comment]
I am sorry I could not find the time to find a good way to handle the `plumed` related issue.

**Comment by junghans** [comment]
@danielahlin FYI, having mpi and non-mpi in one build (like FFTW) would still be nice.

**Comment by eirrgang** [comment]
I don't see a good way to make this happen without either upstream changes to the gromacs build system or a spack virtual package, depending on separate concrete packages for the different build configurations. There are various [outstanding bugs](https://gitlab.com/gromacs/gromacs/-/issues/4334) related to CMake configuration files that cannot coexist in the same installation prefix that are best dealt with by having separate installation prefixes (as is standard for separate concrete specs).

**Comment by danielahlin** [comment]
@junghans - thanks for the ping

Having read through above it is not clear to me that the gains are worth the build complications. IIUC the benefits are limited to having one activation command (i.e. spack load/module add) for (basically PATH-) access to some of the different major variants of gromacs. When/if modules are used, could not the same be achieved by having a super-module loading several sub-modules, one for each variant?

Also apart from complicating the build, this would (again IIUC) come at the expense of the same build settings and dependencies being applied to the major variants, which may be sub-optimal.

Two example to illustrate: 

1. Assume I'm forced to use a particular compiler version to get my preferred MPI implementation to work but that I don't want to use this compiler for the tmpi gmx binary (if I could get better performance with some other compiler). How would this be resolved?

2. With reference to Cray login nodes and login/compute node heterogeneity in general: Many Crays XCs had Ivy Bridge login nodes and Haswell compute nodes. In this case which target wrt SIMD should be set for the common build? With AVX2 they will not run on the login node and without the performance on the compute nodes will be suboptimal. To be clear I'm not arguing XC properties are of particular importance - but rather that in general that the nodes in a particular system sharing a spack deployment may differ significantly and that using the same build settings may not be efficient.

**Comment by eirrgang** [comment]
> Having read through above it is not clear to me that the gains are worth the build complications. IIUC the benefits are limited to having one activation command (i.e. spack load/module add) for (basically PATH-) access to some of the different major variants of gromacs. When/if modules are used, could not the same be achieved by having a super-module loading several sub-modules, one for each variant?

Yes. I don't see that there is necessarily anything in common between the multiple installations (from the builder perspective) other than that they originated from the same source repository. If it is valuable to users to load just one module, I think that should be addressed at a separate level than the build recipe, such as through a virtual/meta package or customization of the module.

**Comment by alalazo** [comment]
Given the comments above, can we close this PR? Or is the matter still debated?

**Comment by alalazo** [comment]
Closing, since there has been no reply in three weeks to https://github.com/spack/spack/pull/23489#issuecomment-1917583984. Feel free to reopen if you want to continue work on this.

**Pull request closed without merging**
